File: bench/closures/closures.go
// Go peer of bench/closures.vader. A captured variable that escapes moves to
// the heap, which is the same cell Vader mints, and a func value is the same
// indirect call.

package main

import "fmt"

const (
	calls  = 6_000_000
	adders = 500
)

func makeAdder(n int64) func(int64) int64 {
	return func(x int64) int64 { return x + n }
}

func main() {
	var counter int64
	bump := func() { counter = counter + 1 }
	for i := 0; i < calls; i++ {
		bump()
	}

	list := make([]func(int64) int64, 0, adders)
	for index := 0; index < adders; index++ {
		list = append(list, makeAdder(int64(index)))
	}
	var total int64
	for pass := 0; pass < calls/adders; pass++ {
		for _, adder := range list {
			total = adder(total)
		}
	}
	fmt.Printf("closures counter=%d total=%d\n", counter, total)
}

File: bench/closures/closures.java
// Java peer of bench/closures.vader. Java cannot mutate a captured local, so
// the counter is a one-element array — what Java code writes, and the same
// pointer hop Vader's heap cell costs.

import java.util.function.LongUnaryOperator;

static final int CALLS = 18_000_000;
static final int ADDERS = 500;

static LongUnaryOperator makeAdder(long n) {
    return x -> x + n;
}

void main() {
    long[] counter = { 0 };
    Runnable bump = () -> counter[0] = counter[0] + 1;
    for (int i = 0; i < CALLS; i++) {
        bump.run();
    }

    LongUnaryOperator[] adders = new LongUnaryOperator[ADDERS];
    for (int index = 0; index < ADDERS; index++) {
        adders[index] = makeAdder(index);
    }
    long total = 0;
    for (int pass = 0; pass < CALLS / ADDERS; pass++) {
        for (LongUnaryOperator adder : adders) {
            total = adder.applyAsLong(total);
        }
    }
    IO.println("closures counter=%d total=%d".formatted(counter[0], total));
}
